### ISSUE TITLE
Memory Leak:  Unable to completely remove a run-time loaded GLB from memory

### ISSUE THREAD

**Issue by Tom-Render** [atteneder/glTFast#726]
First, thanks for the magnificent codebase 😉, this issue has me completely stumped so I’m pretty confident it’s a bug.  However, apologies in advance if it’s not and there’s some elusive means of clearing gLTFast’s memory use I’ve somehow missed in the included example project.  

**Describe the bug**

Every time we load a new GLB model we can only partially remove it from memory.  This seems to happen on every platform, but is particularly problematic for us on iPad, where our application will inevitably build up memory use over time and eventually crash.  To clarify, each one of our models is optimised enough to be viewed individually but viewing them in sequence causes issues because we cannot completely remove their predecessors from RAM.  

**Files**

A small example unity project (including a glb) which re-creates the issue and produces .snap files (unity memory snapshots) as evidence can be downloaded here: https://we.tl/t-jYf1k8LwRZ

**To Reproduce**

1. Download the example project: https://we.tl/t-jYf1k8LwRZ
2. Extract and open it in Unity 2022.3.51f1
3. Open build settings and tick “Development Build”, then do a build
4. Run the built executable, the example application will:
    1. Load an empty scene, then take a first memory snapshot.
    2. Additively load a new scene.
    3. Run-time load a glb inside of that new scene.
    4. Destroy the loaded glb and clear any references to it.
    5. Unload the scene the glb (and the script that loaded it) was in.
    6. Force a memory clean-up with Resources/UnloadUnusedAsset and GC.Collect. 
    7. Take a second memory snapshot
4. Wait until the application says, “Demonstration complete”.  You can then close the application.
5. There will be two new .snap files on your desktop, return to the Unity Editor and use Window -> Analysis -> Memory Profiler to import and compare the .snap files. For example, I get an additional 270mb of untracked memory from the second memory snapshot, when ideally both snap files would have the same memory use. 

**Expected behaviour**

When the first and second memory snapshots are compared, memory usage should be nearly identical.  We should be able to restore RAM usage to (roughly) the original amount it was before the GLB was loaded.

**Screenshots**

![Capture](https://github.com/user-attachments/assets/25089291-8bf6-4a82-8e08-8d0075a798a5)

**Desktop (please complete the following information):**

- glTFast version: 6.8.0 (From the Unity Registry)
- Unity Editor version: 2022.3.51f1
- Render Pipeline and version: Built-In Render Pipeline for 2022.3.51f1
- Platform: Standalone (Windows, Mac, Linux), Windows 10

**Comment by Tom-Render** [comment]
I realised my example above, crucially, doesn’t invoke GltfImport.Dispose, here’s a version of the project which **does** invoke GltfImport.Dispose: https://we.tl/t-dzLqJbvrGf

Sadly, the memory leak is still present even after invoking GltfImport.Dispose:

![Capture](https://github.com/user-attachments/assets/a156d15b-b048-4370-861e-045d91ba2f08)

**Comment by Tom-Render** [comment]
For the sake of completeness, I should also note that I've now tested the example project in Unity 6000.0.25f1 and the leak still occurs.

**Comment by nickconstantinidis** [comment]
I investigated various combinations of the following: 

- Unity Editor versions: 2022.3.X, 6000.0.X, 6000.3.X (latest)
- glTFast versions: 6.8.0, 6.16.0 (latest)
- Load/unload iterations (added loop to provided code): 1-10

In some cases the total memory usage actually decreased between iterations. Generally after 5+ iterations the memory remains stable and no longer climbs. There is a noticeable improvement in `glTFast 6.16.0` compared to `glTFast 6.8.0` as there have been lots of improvements related to native memory usage between those versions.

In all cases the majority (95%+) of the increased memory usage that remains uncleared falls into two categories:

1. Untracked -> Private
2. Native -> Reserved

The former is not directly allocated by Unity, but instead is memory used by the underlying operating system and outside of Unity's control. The latter is likely due to fragmentation and how the native allocators allocate memory in chunks. Both of these are consistent with the observed behaviour that the increases in memory stabilize after multiple iterations and are a result of how Unity and the underlying operation system interact.

Therefore, I would conclude that the behaviour seen here is not out of the ordinary and should not be considered a memory leak.

**Comment by Tom-Render** [comment]
Thanks @nickconstantinidis I really appreciate all the time and expertise you’ve put into investigating this.  

We have an iPad application that downloads models from the internet and views them.  Just to check I’m understanding correctly, your conclusion means the following:

_It's impossible to create an iPad application to view an indefinitely long sequence of different 3D models.  As since the models are different (not just the same model loading repeatedly as per the provided code) sooner or later the amount of increased untracked and native memory will cause a memory overflow and crash the application._    

Or is it just that it’s impossible to create such an application using the Unity engine? 

Sorry if that question is slightly off topic, but I desperately need to make sure I’m correctly understanding the limitations of the technology here so I can pass that knowledge up the chain.

**Comment by nickconstantinidis** [comment]
I don't think that is correct. The maximum amount of used memory should stabilize after loading/unloading enough models. The untracked and native reserved memory are still visible in the profiler but not actively being used. They do eventually get freed up or reused by the application later on when another model is loaded.

It might be possible to load/unload specific models in a particular order in such a way that maximizes the amount of "wasted" memory, that is memory that is still reserved but not usable for a subsequent model to load. But I doubt that would be possible in real-world use cases.

I would instead claim that the amount of total memory required for such an indefinitely long sequence of different 3D models is higher than the amount required for the first model to load, but that it will plateau at some point. There will be a point reached when there is no reason to reserve any new memory since the untracked and reserved memory will be large enough to be reused completely.

Your theoretical application should indeed be possible, although the amount of memory required might be higher than it initially seems. In practice, it might be possible to determine what this limit is by iterating through load/unload sequences until the memory usage stops rising altogether. I'd estimate that somewhere between 10 and 20 iterations would suffice to find this limit, but that may depend on the models used for this test. I myself admittedly only tested this with the single jet model in the project you uploaded.

**Comment by Tom-Render** [comment]
Thanks @nickconstantinidis that's a relief, and very well explained.  

I'll set up a new project with a long range of models and get some hard data to make sure.  But it'll be a few weeks before my schedule clears up enough for that.  If the issue is still open come then, I'll post the results here.  I fully expect you'll be correct, but it makes sense to get some closure.
